synapseclient/client.py

```diff
@@ -621,6 +621,72 @@ def invalidateAPIKey(self):
         if self._is_logged_in():
             self.restDELETE("/secretKey", endpoint=self.authEndpoint)
 
+    @functools.lru_cache()
+    def get_user_profile_by_username(
+        self,
+        id: str = None,
+        sessionToken: str = None,
+    ) -> UserProfile:
+        """
+        Get the details about a Synapse user.
+        Retrieves information on the current user if 'id' is omitted or is empty string.
+        :param id:           The userName of a user
+        :param sessionToken: The session token to use to find the user profile
+        :returns: The user profile for the user of interest.
+
+        Example::
+            my_profile = syn.get_user_name_profile()
+            freds_profile = syn.get_user_name_profile('fredcommo')
+        """
+        if id:
+            if isinstance(id, str):
+                principals = self._findPrincipals(id)
+                for principal in principals:
+                    if principal.get("userName", None).lower() == id.lower():
+                        id = principal["ownerId"]
+                        break
+                else:
+                    raise ValueError(f"Can't find user '{id}'")
+            else:
+                raise TypeError("id must be a 'userName' string")
+        else:
+            id = ""
```

Comment:
Can there be a way to un-nest this?  For example, did we want to fast fail if the username isn't a string or None?

Reply:
I moved the `TypeError` to the top of the function in the latest commit